### ISSUE TITLE
Declaration inside conditional

### ISSUE THREAD

**Issue by eevleevs** [DanielXMoore/Civet#444]
I would use often if 

```js
if a := 1
  something
```

translated to something like

```js
{
  const a = 1
  if (a) {
    something
  }
}
```

all combinations with `let`, `while`, `unless`, etc. also welcome.

**Comment by edemaine** [comment]
Agreed, this would be super useful; we've discussed it on Discord before.  In particular, I'd use this a lot with regular expressions (`if match := regex.exec ...`).

A more likely compilation for your example would be:

```js
let a
if (a = 1) {
  something
}
```

This is useful because it preserves the order of operations.  For example, if you had the expression `f() + (a := g())`, we'd expect `f()` to be called before `g()`, so the assignment should happen within the expression instead of in the declaration.

To preserve the `const` nature (no reassignment), @STRd6 has proposed this sort of compilation:

```js
let a1
if (a1 = 1) {
  const a = a1
  something
}
```

In addition to implementing correct `const` semantics (except within the `if` clause itself, where I guess we'd need to check for a single assignment), this would make `a` accessible only within the `if` block, not in an `else` block or outside the `if...else` clause.  It's similar to the behavior of `for (const ...)`, which scopes the variable to exist only within the `for` loop.

**Comment by STRd6** [comment]
This could also work with destructuring inside the conditional:

```coffee
if [a, b, c] := x
  a + b + c
```

```typescript
let ref
if (ref = x) {
  const [a, b, c] = ref
  a + b + c
}
```

**Comment by edemaine** [comment]
That would be super useful!  The number of times I've assigned `match` from a regular expression execution, checked for it being defined, only to immediately destructure it and ignore `match`...

**Comment by STRd6** [comment]
Fixed in #451
